### ISSUE TITLE
Allow using underscore class names for ActionCable channel names

### ISSUE THREAD

**Issue by TikiTDO** [rails/rails#25196]
In the current version of ActionCable creating a connection from the browser to the server requires code like this:

``` javascript
  App.cable.subscriptions.create("ExampleChannel")
```

It would be nice to use underscored names, such as: 

``` javascript
  App.cable.subscriptions.create("example_channel")
```

---

This can be done by modifying [this line](https://github.com/rails/rails/blob/52ce6ece8c8f74064bb64e0a0b1ddd83092718e1/actioncable/lib/action_cable/connection/subscriptions.rb#L29) to something like:

``` ruby
  subscription_klass = connection.server.channel_classes[id_options[:channel].camelize]
```

---

Alternatively, the code [here](https://github.com/rails/rails/blob/5ec889eb8b81be1578a30373dacb57c765700157/actioncable/lib/action_cable/server/base.rb#L73) could be modified to to key they modules by both camel case and underscore names as such:

``` ruby
  @channel_classes ||= begin
    config.channel_paths.each { |channel_path| require channel_path }
    config.channel_class_names.each_with_object({}) do |name, hash| 
      hash[name] = name.constantize
      hash[name.underscore] = name.constantize
    end
  end
```

**Comment by maclover7** [comment]
cc @javan

**Comment by jeremy** [comment]
> It would be nice to use underscored names, such as:

Could you talk about why?

**Comment by TikiTDO** [comment]
Primarily I'm looking for consistency. Most of the times when client side code is referring to server side resources it uses underscores, as with routes. This becomes even more noticeable if you want to use namespaced channels: ``Namespace::ExampleChannel` vs `namespace/example_channel`.

At some point it honestly just feels wrong to use full ruby class names in javascript/coffeescript. The resulting code doesn't fit very well with the conventions of either language, nor with the conventions rails has established over the past decade. 

Beyond philosophical reasons, my personal usecase is a bit peculiar. I am writing my own ActionCable JS adapter to fit into my own connection management framework, which in turn allows me to do something like `server.example_channel.example_action(params...)`. In this case having underscored channel names would offer consistent naming between actions and channels, which in turn more closely resembles routes.

Normally I would just throw in an initializer to patch in the functionality I'm looking for at leave it at that, but upon thinking about it I figured that this is a feature that could be useful to others.

**Comment by jeremy** [comment]
> client side code is referring to server side resources
> […]
> just feels wrong to use full ruby class names

This is creating a channel class on the client. It is a close ally to the Ruby channel class, but we'd use camel-case naming if this were JS-only, too.

> do something like `server.example_channel.example_action(params...)`

You can do this with a camelize/underscore convention analogous to Rails controllers and routes!

**Comment by TikiTDO** [comment]
> This is creating a channel class on the client. It is a close ally to the Ruby channel class, but we'd use camel-case naming if this were JS-only, too.

Technically it is creating an instance of the [ActionCable.Subscription](https://github.com/rails/rails/blob/master/actioncable/app/assets/javascripts/action_cable/subscription.coffee) class, which then uses the identifier whenever a request is sent or received. Certainly the two are related, but I would say they are related in the same way that a route to `"example#action"` is related to `ExampleController#action`.

Pretty much every time there is this level of coupling in Rails (routes, AR relation names, names of view directories) it seems to use the underscore convention. In all honestly, then only time I have seen CamelCase used in this sort of context is the class_name attribute of relations, where it's referring directly to a ruby class.

> You can do this with a camelize/underscore convention analogous to Rails controllers and routes!

I would need to configure my JS underscore equivalent with all the same ruby inflections, or risk unexpected problems with acronyms, and possibly other inflections.

**Comment by javan** [comment]
I don't see a compelling reason to allow `subscriptions.create("example_channel")` *unless* we drop the `channel` part.  Map `subscriptions.create("example")` to `ExampleChannel`, `subscriptions.create("notifications/web_alert")` to `Notifications::WebAlertChannel`, etc.

**Comment by TikiTDO** [comment]
That would be even better to be honest, in that case the entire thing becomes almost completely analogous to routes, which suits my use case perfectly.

**Comment by TikiTDO** [comment]
I was trying to implement something that would drop the channel part as @javan proposed. I made the following changes to [this location](https://github.com/rails/rails/blob/5ec889eb8b81be1578a30373dacb57c765700157/actioncable/lib/action_cable/server/base.rb#L73)

```ruby
  @channel_classes ||= begin
    config.channel_paths.each { |channel_path| require channel_path }
    config.channel_class_names.each_with_object({}) do |name, hash| 
      hash[name] = name.constantize
      hash[name.underscore.gsub(/_channel$/, '')] = name.constantize
    end
  end
```

I was able to test that `subscriptions.create("example")` and `subscriptions.create("ExampleChannel")` both work, and can properly initiate communication from both server and client.

I could not test namespacing, because I ran into #25224 which prevents namespacing from working in channels all together.
